stripe/src/main/java/com/stripe/android/paymentsheet/ui/PrimaryButton.kt

```diff
@@ -2,13 +2,103 @@ package com.stripe.android.paymentsheet.ui
 
 import android.content.Context
 import android.util.AttributeSet
+import android.view.LayoutInflater
 import android.widget.FrameLayout
+import androidx.core.view.isVisible
+import com.stripe.android.R
+import com.stripe.android.databinding.PrimaryButtonBinding
 
 /**
  * The primary call-to-action for a payment sheet screen.
  */
-internal abstract class PrimaryButton @JvmOverloads constructor(
+internal class PrimaryButton @JvmOverloads constructor(
     context: Context,
     attrs: AttributeSet? = null,
     defStyleAttr: Int = 0
-) : FrameLayout(context, attrs, defStyleAttr)
+) : FrameLayout(context, attrs, defStyleAttr) {
+
+    private var state: State? = null
+    private val animator = PrimaryButtonAnimator(context)
+
+    internal val viewBinding = PrimaryButtonBinding.inflate(
+        LayoutInflater.from(context),
+        this
+    )
+
+    private val confirmedIcon = viewBinding.confirmedIcon
+
+    init {
+        setBackgroundResource(R.drawable.stripe_paymentsheet_primary_button_default_background)
+
+        isClickable = true
+        isEnabled = false
+    }
+
+    fun setLabelText(text: String) {
+        viewBinding.label.text = text
+    }
+
+    private fun onReadyState(text: String) {
+        setLabelText(text)
+        viewBinding.confirmingIcon.isVisible = false
+    }
+
+    private fun onConfirmState() {
+        viewBinding.lockIcon.isVisible = false
+        viewBinding.confirmingIcon.isVisible = true
+
+        viewBinding.label.text = resources.getString(
+            R.string.stripe_paymentsheet_button_processing
+        )
+    }
+
+    private fun onCompletedState(onAnimationEnd: () -> Unit) {
+        setBackgroundResource(
+            R.drawable.stripe_paymentsheet_primary_button_confirmed_background
+        )
+
+        animator.fadeOut(viewBinding.label)
+        animator.fadeOut(viewBinding.confirmingIcon)
+
+        animator.fadeIn(confirmedIcon, width) {
+            onAnimationEnd()
+        }
+    }
+
+    override fun setEnabled(enabled: Boolean) {
+        super.setEnabled(enabled)
+        viewBinding.lockIcon.isVisible = enabled
+        updateAlpha()
+    }
+
+    fun updateState(state: State?, completeCallback: () -> Unit = {}) {
```

Comment:
Can we make `completeCallback` a property on `PrimaryButton` instead of an argument of `updateState`?

Reply:
I think reading it in the calling code it is clearer to send it in here because you know exactly that there is a callback and what will happen as a result, you don't have to go looking in another place.

Reply:
I think the method signature is a little confusing for a future maintainer. `completeCallback` is accepted for all `State` types, but is only passed through for the `Completed` state.